### PR TITLE
feat(AI Transform Node): New node

### PR DESCRIPTION
**Pull request by michael-radency**
## Summary

This PR introduces a new "AI Transform"* node that generates code snippets based on user instructions. The AI is context-aware, understanding the workflow's nodes and their data types. Users can simply enter their instructions and click the "Generate Code" button to let the AI automatically populate the Code parameter.

image

---
During development, the node's "button" property was improved:
```
export type NodePropertyAction = {
	type: 'askAiCodeGeneration';
	handler?: string;
	target?: string;
};

export interface INodePropertyTypeOptions {
	
	buttonConfig?: {
		action?: string | NodePropertyAction;
		label?: string; 
		hasInputField?: boolean;
		inputFieldMaxLength?: number; 
	};
	...
}
```
This updated interface allows you to specify a label, an optional input field with its maximum input length, and an action. The action can be specified as a string or as a `NodePropertyAction`, which allows you to set the type of action (currently only 'askAiCodeGeneration') and optionally an action target, such as a parameter that would be affected as a result of this action. Additionally, a handler function defined in `methods.actionHandler` can be executed on the backend when clicked.
```
	methods = {
		actionHandler: {
			async generateCode(this: ILoadOptionsFunctions, payload: IDataObject | string | undefined) {
				const url = this.getNodeParameter('url', 0) as string;
				let body: IDataObject = {};
				if (payload) {
					body = typeof payload === 'string' ? jsonParse(payload) : payload;
				}
				const response = (await this.helpers.httpRequest({
					method: 'POST',
					url,
					headers: {
						'Content-Type': 'application/json',
					},
					body,
				})) as {
					data: { code: string };
				};

				return response?.data?.code;
			},
		},
	};
```
---
* This node is currently available only on the cloud.

## Related Linear tickets, Github issues, and Community forum posts
https://linear.app/n8n/issue/NODE-1461/ai-transform-node-p0

### CODE REVIEW

File: packages/editor-ui/src/components/ButtonParameter.vue [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+}
+
+function stopLoading() {
+	setTimeout(() => {
+		isLoading.value = false;
+	}, 200);
+}
+
+function getPath(parameter: string) {
+	return ((props.path ? `${props.path}.` : '') + parameter) as string;
```
Comment (ShireenMissi) [inline packages/editor-ui/src/components/ButtonParameter.vue:89]:
```suggestion
	return ((props.path ? `${props.path}.` : '') + parameter);
```

File: packages/editor-ui/src/components/ButtonParameter.vue [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+	if (!activeNode || !workflow) return [];
+
+	return workflow
+		.getParentNodesByDepth(activeNode?.name)
+		.filter(({ name }, i, nodes) => {
+			return name !== activeNode.name && nodes.findIndex((node) => node.name === name) === i;
+		})
+		.map((n) => getNodeByName(n.name))
+		.filter((n) => n !== null) as INodeUi[];
```
Comment (ShireenMissi) [inline packages/editor-ui/src/components/ButtonParameter.vue:75]:
```suggestion
		.filter((n) => n !== null);
```

File: packages/editor-ui/src/components/ButtonParameter.vue [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+		</div>
+		<div :class="$style.controls">
+			<N8nTooltip :disabled="isSubmitEnabled">
+				<div>
+					<N8nButton
+						:disabled="!isSubmitEnabled"
+						size="small"
+						@click="onSubmit"
+						:loading="isLoading"
+						type="secondary"
```
Comment (ShireenMissi) [inline packages/editor-ui/src/components/ButtonParameter.vue:253]:
```suggestion
					        :loading="isLoading"
						type="secondary"
						@click="onSubmit"
```

File: packages/editor-ui/src/components/CodeNodeEditor/utils.ts [hunk 4]
```diff
@@ -53,3 +59,43 @@ export const addInfoRenderer = (option: Completion): Completion => {
...
+
+	if (!activeNode || !workflow) return [];
+
+	return workflow
+		.getParentNodesByDepth(activeNode?.name)
+		.filter(({ name }, i, nodes) => {
+			return name !== activeNode.name && nodes.findIndex((node) => node.name === name) === i;
+		})
+		.map((n) => getNodeByName(n.name))
+		.filter((n) => n !== null) as INodeUi[];
```
Comment (ShireenMissi) [inline packages/editor-ui/src/components/CodeNodeEditor/utils.ts:76]:
```suggestion
		.filter((n) => n !== null);
```

File: packages/nodes-base/nodes/AiTransform/AiTransform.node.json [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+	"details": "Modify data by writing a prompt",
+	"categories": ["Development", "Core Nodes"],
+	"resources": {
+		"primaryDocumentation": [
+			{
+				"url": "https://docs.n8n.io/integrations/builtin/core-nodes/n8n-nodes-base.aitransform/"
+			}
+		]
+	},
+	"alias": ["cpde", "Javascript", "JS", "Script", "Custom Code", "Function", "AI", "LLM"],
```
Comment (ShireenMissi) [inline packages/nodes-base/nodes/AiTransform/AiTransform.node.json:14]: Do you mean code?
```suggestion
	"alias": ["code", "Javascript", "JS", "Script", "Custom Code", "Function", "AI", "LLM"],
```

**Review by ShireenMissi — COMMENTED** [review]
Mostly looks good 👍, left a few lint comments 
Please Add tests and update the PR description with more information about the changes

File: packages/editor-ui/src/stores/nodeTypes.store.ts [hunk 6]
```diff
@@ -321,6 +327,7 @@ export const useNodeTypesStore = defineStore(STORES.NODE_TYPES, () => {
 		visibleNodeTypesByInputConnectionTypeNames,
 		isConfigurableNode,
 		getResourceMapperFields,
+		getNodeParameterActionResult,
```
[the remark this answers is not in the document]
Comment (michael-radency) [reply packages/editor-ui/src/stores/nodeTypes.store.ts:330]: not currently, but this is part of button parameter where you could use handler assigned in buttonConfig similar to loadOptions, usage example [here](https://github.com/n8n-io/n8n/pull/9990/commits/f3f4c9cc00ad2b98b6a828d1f818d809f4cff335)

**Review by ShireenMissi — APPROVED** [review]
LGTM 🎉

**Comment by janober** [comment]
Got released with [`n8n@1.55.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.55.0)

**Pull request merged**